### ISSUE TITLE
v1.4.50: RCS1198 "Avoid unnecessary boxing of value type" incorrectly triggered by interpolated strings

### ISSUE THREAD

**Issue by IanKemp** [JosefPihrt/Roslynator#105]
```csharp
var b = byte.MaxValue;
var s = $"foo { b } bar"; // b triggers RCS1198
```

If written directly as `string.Format` (which Roslyn internally converts interpolated strings to), no warning is generated:

```csharp
var b = byte.MaxValue;
var s = string.Format("foo {0} bar", b); // no RCS1198
```

**Comment by IanKemp** [comment]
Also occurs with ordinary string concatenation in some cases:

```
private IEnumerable<object> ToObjectArray(Array array, int max)
{
    for (var i = 0; i < Math.Min(array.Length, max); i++)
    {
        yield return array.GetValue(i);
    }

    if (array.Length > max)
    {
        yield return "...x" + array.Length; // array.Length erroneously triggers RCS1198
    }
}
```

**Comment by JosefPihrt** [comment]
Diagnostic is reasonable in this case. You can see below that there is a boxing.

### Without 'ToString'
```csharp
private static void Main(string[] args)
{
    string s = "x" + args.Length;
}
```
```
.method private hidebysig static void  Main(string[] args) cil managed
{
  .entrypoint
  // Code size       20 (0x14)
  .maxstack  8
  IL_0000:  ldstr      "...x"
  IL_0005:  ldarg.0
  IL_0006:  ldlen
  IL_0007:  conv.i4
  IL_0008:  box        [System.Runtime]System.Int32
  IL_000d:  call       string [System.Runtime]System.String::Concat(object,
                                                                    object)
  IL_0012:  pop
  IL_0013:  ret
} // end of method Program::Main
```

### With 'ToString'
```csharp
private static void Main(string[] args)
{
    string s = "x" + args.Length.ToString();
}
```
```
.method private hidebysig static void  Main(string[] args) cil managed
{
  .entrypoint
  // Code size       23 (0x17)
  .maxstack  2
  .locals init (int32 V_0)
  IL_0000:  ldstr      "...x"
  IL_0005:  ldarg.0
  IL_0006:  ldlen
  IL_0007:  conv.i4
  IL_0008:  stloc.0
  IL_0009:  ldloca.s   V_0
  IL_000b:  call       instance string [System.Runtime]System.Int32::ToString()
  IL_0010:  call       string [System.Runtime]System.String::Concat(string,
                                                                    string)
  IL_0015:  pop
  IL_0016:  ret
} // end of method Program::Main
```

**Comment by IanKemp** [comment]
My apologies - you are completely correct, and in fact according to the IL I have looked at, `string.Format` **does** also perform boxing, so in theory it should also trigger this diagnostic. So this bug report is somewhat the wrong way around. :)

[But as has been pointed out](https://stackoverflow.com/a/8477372/70345), when it comes to string formatting, boxing is almost certainly not going to be the biggest performance issue. (I would guess that this is also true for concatenation, although less so.) As such, this diagnostic is IMO superfluous for strings and just creates noise for them (this is the main reason I'm posting this report: I've disabled RCS1198 because 99% of what it complains about are strings, but I'm worried about the 1% of non-string code I'm going to miss as a result).

There is also another issue with this diagnostic, in that it's triggered regardless of whether you're using format characters in strings - in that case you **have** to use boxing to get the expected result. Consider:

```
        static void Main(string[] args)
        {
            var d = 1234.5678m;
            var s = $"{d:n6}"; // d triggers RCS1198
            Console.WriteLine(s); // CORRECT: 1 234.567800 (on my machine at least)
            Console.ReadKey(true);
        }
```

vs

```
        static void Main(string[] args)
        {
            var d = 1234.5678m;
            var s = $"{d.ToString():n6}"; // no RCS1198
            Console.WriteLine(s); // WRONG: 1234.5678
            Console.ReadKey(true);
        }
```

In summary, I think this would be best:

1. Correct RCS1198 so it also triggers for `string.Format` (essentially the topic of this bug report).
2. Correct RCS1198 so it does not trigger for string interpolation and/or `string.Format` when format strings are involved (can open a new bug for this if you wish). I have a feeling this one will be tricky though.
3. Split RCS1198 into two different analyzers - one for everything but strings, one for strings only - and have the former enabled by default, with the latter disabled. This definitely falls into feature request territory.

**Comment by JosefPihrt** [comment]
I'm not clear on point 3. Could you clarify what is meant by "everything but strings" and "strings only"?

**Comment by IanKemp** [comment]
Sorry - I meant to write "one for everything but **string formatting/concatenation**, one for **string **formatting/concatenation** only'.

**Comment by JosefPihrt** [comment]
I agree that this analyzer should be disabled by default, since it can be perceived as more annoying than helpful.

I've realized that `format` can be variable and thus it is not possible to check it if it contains alignment and/ format strings. As a result:
* I will remove `StringBuilder.AppendFormat` from this analyzer.
* It does not make sense to add analysis of `string.Format` calls.

I'm not sure if it makes sense to create another analyzer. Currently this analyzer covers following syntaxes:
* `StringBuilder.Append(object)`
* `StringBuilder.Insert(int, object)`
* `StringBuilder.AppendFormat` (will be removed)
* string concatenation
* interpolation
